File: android/src/main/kotlin/com/example/zendesk_sdk/commands/ZendeskCommand.kt
package com.example.zendesk_sdk.commands

import com.example.zendesk_sdk.ZendeskSdkChannel
import com.example.zendesk_sdk.models.ZendeskConfig
import com.example.zendesk_sdk.models.ZendeskTicketRequest
import com.example.zendesk_sdk.models.ZendeskUser
import io.flutter.plugin.common.MethodCall

/// Command contract mirroring Dart [ZendeskCommand] (Command pattern + OCP).
sealed interface ZendeskCommand {
    val methodName: String
}

data class InitializeCommand(
    val config: ZendeskConfig,
    val user: ZendeskUser,
) : ZendeskCommand {
    override val methodName: String = ZendeskSdkChannel.Method.INITIALIZE
}

data object LogoutCommand : ZendeskCommand {
    override val methodName: String = ZendeskSdkChannel.Method.LOGOUT
}

data object IsInitializedCommand : ZendeskCommand {
    override val methodName: String = ZendeskSdkChannel.Method.IS_INITIALIZED
}

data class ShowHelpCenterCommand(
    val categoryIdList: List<Long>,
) : ZendeskCommand {
    override val methodName: String = ZendeskSdkChannel.Method.SHOW_HELP_CENTER
}

data class ShowHelpCenterArticleIdCommand(
    val articleId: Long,
) : ZendeskCommand {
    override val methodName: String = ZendeskSdkChannel.Method.SHOW_HELP_CENTER_ARTICLE_ID
}

data class ShowHelpCenterCategoryIdCommand(
    val categoryId: Long,
) : ZendeskCommand {
    override val methodName: String = ZendeskSdkChannel.Method.SHOW_HELP_CENTER_CATEGORY_ID
}

data class SendUserInformationForTicketCommand(
    val request: ZendeskTicketRequest,
) : ZendeskCommand {
    override val methodName: String = ZendeskSdkChannel.Method.SEND_USER_INFORMATION_FOR_TICKET
}

data object ShowListOfTicketsCommand : ZendeskCommand {
    override val methodName: String = ZendeskSdkChannel.Method.SHOW_LIST_OF_TICKETS
}

data class StartChatCommand(
    val channelId: String,
) : ZendeskCommand {
    override val methodName: String = ZendeskSdkChannel.Method.START_CHAT
}

data object StartChatBotCommand : ZendeskCommand {
    override val methodName: String = ZendeskSdkChannel.Method.START_CHAT_BOT
}

data object GetUnreadMessageCountCommand : ZendeskCommand {
    override val methodName: String = ZendeskSdkChannel.Method.GET_UNREAD_MESSAGE_COUNT
}

data class UpdatePushNotificationTokenCommand(
    val token: String,
) : ZendeskCommand {
    override val methodName: String = ZendeskSdkChannel.Method.UPDATE_PUSH_NOTIFICATION_TOKEN
}

data class HandlePushNotificationCommand(
    val payload: Map<String, String>,
) : ZendeskCommand {
    override val methodName: String = ZendeskSdkChannel.Method.HANDLE_PUSH_NOTIFICATION
}

sealed class ZendeskCommandParseResult {
    data class Success(val command: ZendeskCommand) : ZendeskCommandParseResult()
    data object InvalidArguments : ZendeskCommandParseResult()
    data object NotImplemented : ZendeskCommandParseResult()
}

object ZendeskCommandFactory {
    fun parse(call: MethodCall): ZendeskCommandParseResult {
        return when (call.method) {
            ZendeskSdkChannel.Method.INITIALIZE -> {
                val config = ZendeskConfig.fromMethodCall(call)
                    ?: return ZendeskCommandParseResult.InvalidArguments
                ZendeskCommandParseResult.Success(
                    InitializeCommand(config = config, user = ZendeskUser.fromMethodCall(call)),
                )
            }

            ZendeskSdkChannel.Method.LOGOUT ->
                ZendeskCommandParseResult.Success(LogoutCommand)

            ZendeskSdkChannel.Method.IS_INITIALIZED ->
                ZendeskCommandParseResult.Success(IsInitializedCommand)

            ZendeskSdkChannel.Method.SHOW_HELP_CENTER -> {
                val categoryIdList =
                    call.argument<List<Long>>(ZendeskSdkChannel.Argument.CATEGORY_ID_LIST)
                        ?: emptyList()
                ZendeskCommandParseResult.Success(
                    ShowHelpCenterCommand(categoryIdList = categoryIdList),
                )
            }

            ZendeskSdkChannel.Method.SHOW_HELP_CENTER_ARTICLE_ID -> {
                val articleId = call.argument<String>(ZendeskSdkChannel.Argument.ARTICLE_ID)
                    ?.toLongOrNull()
                    ?: return ZendeskCommandParseResult.InvalidArguments
                ZendeskCommandParseResult.Success(
                    ShowHelpCenterArticleIdCommand(articleId = articleId),
                )
            }

            ZendeskSdkChannel.Method.SHOW_HELP_CENTER_CATEGORY_ID -> {
                val categoryId = call.argument<String>(ZendeskSdkChannel.Argument.CATEGORY_ID)
                    ?.toLongOrNull()
                    ?: return ZendeskCommandParseResult.InvalidArguments
                ZendeskCommandParseResult.Success(
                    ShowHelpCenterCategoryIdCommand(categoryId = categoryId),
                )
            }

            ZendeskSdkChannel.Method.SEND_USER_INFORMATION_FOR_TICKET -> {
                val request = ZendeskTicketRequest.fromMethodCall(call)
                    ?: return ZendeskCommandParseResult.InvalidArguments
                ZendeskCommandParseResult.Success(
                    SendUserInformationForTicketCommand(request = request),
                )
            }

            ZendeskSdkChannel.Method.SHOW_LIST_OF_TICKETS ->
                ZendeskCommandParseResult.Success(ShowListOfTicketsCommand)

            ZendeskSdkChannel.Method.START_CHAT -> {
                val channelId = call.argument<String>(ZendeskSdkChannel.Argument.CHANNEL_ID)
                if (channelId.isNullOrBlank()) {
                    return ZendeskCommandParseResult.InvalidArguments
                }
                ZendeskCommandParseResult.Success(StartChatCommand(channelId = channelId))
            }

            ZendeskSdkChannel.Method.START_CHAT_BOT ->
                ZendeskCommandParseResult.Success(StartChatBotCommand)

            ZendeskSdkChannel.Method.GET_UNREAD_MESSAGE_COUNT ->
                ZendeskCommandParseResult.Success(GetUnreadMessageCountCommand)

            ZendeskSdkChannel.Method.UPDATE_PUSH_NOTIFICATION_TOKEN -> {
                val token = call.argument<String>(ZendeskSdkChannel.Argument.PUSH_TOKEN)
                if (token.isNullOrBlank()) {
                    return ZendeskCommandParseResult.InvalidArguments
                }
                ZendeskCommandParseResult.Success(
                    UpdatePushNotificationTokenCommand(token = token),
                )
            }

            ZendeskSdkChannel.Method.HANDLE_PUSH_NOTIFICATION -> {
                @Suppress("UNCHECKED_CAST")
                val payload = call.argument<Map<String, Any>>(
                    ZendeskSdkChannel.Argument.PUSH_NOTIFICATION_DATA,
                ) ?: emptyMap()
                ZendeskCommandParseResult.Success(
                    HandlePushNotificationCommand(
                        payload = payload.mapValues { it.value.toString() },
                    ),
                )
            }

            else -> ZendeskCommandParseResult.NotImplemented
        }
    }
}

File: ios/zendesk_sdk/Sources/zendesk_sdk/Commands/ZendeskCommand.swift
import Flutter
import Foundation

/// Command contract mirroring Dart ZendeskCommand (Command pattern + OCP).
protocol ZendeskCommand {
    var methodName: String { get }
}

struct InitializeCommand: ZendeskCommand {
    let config: ZendeskConfig
    let user: ZendeskUser
    var methodName: String { ZendeskSdkChannel.Method.initialize }
}

struct LogoutCommand: ZendeskCommand {
    var methodName: String { ZendeskSdkChannel.Method.logout }
}

struct IsInitializedCommand: ZendeskCommand {
    var methodName: String { ZendeskSdkChannel.Method.isInitialized }
}

struct ShowHelpCenterCommand: ZendeskCommand {
    let categoryIdList: [NSNumber]
    var methodName: String { ZendeskSdkChannel.Method.showHelpCenter }
}

struct ShowHelpCenterArticleIdCommand: ZendeskCommand {
    let articleId: String
    var methodName: String { ZendeskSdkChannel.Method.showHelpCenterArticleId }
}

struct ShowHelpCenterCategoryIdCommand: ZendeskCommand {
    let categoryId: Int64
    var methodName: String { ZendeskSdkChannel.Method.showHelpCenterCategoryId }
}

struct SendUserInformationForTicketCommand: ZendeskCommand {
    let request: ZendeskTicketRequest
    var methodName: String { ZendeskSdkChannel.Method.sendUserInformationForTicket }
}

struct ShowListOfTicketsCommand: ZendeskCommand {
    var methodName: String { ZendeskSdkChannel.Method.showListOfTickets }
}

struct StartChatCommand: ZendeskCommand {
    let channelId: String
    var methodName: String { ZendeskSdkChannel.Method.startChat }
}

struct StartChatBotCommand: ZendeskCommand {
    var methodName: String { ZendeskSdkChannel.Method.startChatBot }
}

struct GetUnreadMessageCountCommand: ZendeskCommand {
    var methodName: String { ZendeskSdkChannel.Method.getUnreadMessageCount }
}

struct UpdatePushNotificationTokenCommand: ZendeskCommand {
    let token: String
    var methodName: String { ZendeskSdkChannel.Method.updatePushNotificationToken }
}

struct HandlePushNotificationCommand: ZendeskCommand {
    let payload: [String: Any]
    var methodName: String { ZendeskSdkChannel.Method.handlePushNotification }
}

enum ZendeskCommandParseResult {
    case success(ZendeskCommand)
    case invalidArguments
    case notImplemented
}

enum ZendeskCommandFactory {
    static func parse(call: FlutterMethodCall) -> ZendeskCommandParseResult {
        switch call.method {
        case ZendeskSdkChannel.Method.initialize:
            guard let config = ZendeskConfig.from(call: call) else {
                return .invalidArguments
            }
            return .success(InitializeCommand(config: config, user: ZendeskUser.from(call: call)))

        case ZendeskSdkChannel.Method.logout:
            return .success(LogoutCommand())

        case ZendeskSdkChannel.Method.isInitialized:
            return .success(IsInitializedCommand())

        case ZendeskSdkChannel.Method.showHelpCenter:
            guard let args = call.arguments as? [String: Any],
                  let categoryIdList = args[ZendeskSdkChannel.Argument.categoryIdList] as? [NSNumber] else {
                return .invalidArguments
            }
            return .success(ShowHelpCenterCommand(categoryIdList: categoryIdList))

        case ZendeskSdkChannel.Method.showHelpCenterArticleId:
            guard let args = call.arguments as? [String: Any],
                  let articleId = args[ZendeskSdkChannel.Argument.articleId] as? String,
                  !articleId.isEmpty,
                  Int64(articleId) != nil else {
                return .invalidArguments
            }
            return .success(ShowHelpCenterArticleIdCommand(articleId: articleId))

        case ZendeskSdkChannel.Method.showHelpCenterCategoryId:
            guard let args = call.arguments as? [String: Any],
                  let categoryIdRaw = args[ZendeskSdkChannel.Argument.categoryId] as? String,
                  let categoryId = Int64(categoryIdRaw) else {
                return .invalidArguments
            }
            return .success(ShowHelpCenterCategoryIdCommand(categoryId: categoryId))

        case ZendeskSdkChannel.Method.sendUserInformationForTicket:
            guard let request = ZendeskTicketRequest.from(call: call) else {
                return .invalidArguments
            }
            return .success(SendUserInformationForTicketCommand(request: request))

        case ZendeskSdkChannel.Method.showListOfTickets:
            return .success(ShowListOfTicketsCommand())

        case ZendeskSdkChannel.Method.startChat:
            guard let args = call.arguments as? [String: Any],
                  let channelId = args[ZendeskSdkChannel.Argument.channelId] as? String,
                  !channelId.isEmpty else {
                return .invalidArguments
            }
            return .success(StartChatCommand(channelId: channelId))

        case ZendeskSdkChannel.Method.startChatBot:
            return .success(StartChatBotCommand())

        case ZendeskSdkChannel.Method.getUnreadMessageCount:
            return .success(GetUnreadMessageCountCommand())

        case ZendeskSdkChannel.Method.updatePushNotificationToken:
            guard let args = call.arguments as? [String: Any],
                  let token = args[ZendeskSdkChannel.Argument.pushToken] as? String,
                  !token.isEmpty else {
                return .invalidArguments
            }
            return .success(UpdatePushNotificationTokenCommand(token: token))

        case ZendeskSdkChannel.Method.handlePushNotification:
            guard let args = call.arguments as? [String: Any],
                  let payload = args[ZendeskSdkChannel.Argument.pushNotificationData] as? [String: Any] else {
                return .invalidArguments
            }
            return .success(HandlePushNotificationCommand(payload: payload))

        default:
            return .notImplemented
        }
    }
}
